### PR TITLE
Optional Rate Limiting for BufferingSubscriber

### PR DESCRIPTION
**Pull request by 0aix**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: src/main/java/com/yahoo/bullet/pubsub/BufferingSubscriber.java [hunk 1]
```diff
@@ -22,13 +23,28 @@
...
 
+    /**
+     * The maximum number of PubSubMessages we can receive in a rate limit interval. Further calls to receive will return nothing.
+     */
+    protected final int rateLimitMaxMessages;
+
+    /**
+     * The duration of a rate limit interval.
+     */
+    protected final long rateLimitIntervalMs;
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/pubsub/BufferingSubscriber.java:41]: MS

File: src/test/java/com/yahoo/bullet/pubsub/BufferingSubscriberTest.java [hunk 2]
```diff
@@ -66,6 +72,26 @@ public void testMaxUncommittedMessages() throws PubSubException {
...
+        List<PubSubMessage> messages = make(20);
+        ExampleBufferingSubscriber subscriber = new ExampleBufferingSubscriber(100, 5, 1000L, messages);
+        for (int i = 0; i < 5; i++) {
+            Assert.assertNotNull(subscriber.receive());
+        }
+        Assert.assertNull(subscriber.receive());
+        Assert.assertEquals(subscriber.getCallCount(), 5);
+
+        // Sleep to reset interval
+        Thread.sleep(1500);
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/pubsub/BufferingSubscriberTest.java:86]: Let's sleep for 1-10 ms instead of 1.5s

**Review by akshaisarma — COMMENTED** [review]
Coverage dropped?

**Comment by 0aix** [comment]
> Coverage dropped?

Maybe because I added a lombok constructor..? I have no clue

**Review by akshaisarma — APPROVED** [review]

**Pull request merged**